File: website/static/scripts/api/user-data/flashcards.js
import { request } from "../client.js";

/**
 * Gets the review times for the different ratings a user can give to a word
 * 
 * @param {string} word Word to get next review times for
 */
export async function fetchReviewTimes(word) {
    return await request('/fetch_review_times', 'POST', word);
}

/**
 * Gets the next word and card for the user to review
 * 
 * @param {number} [wordIndex=0] Index of list of due words to use
 */
export async function fetchNextWord(wordIndex = 0) {
    return await request('/fetch_next_word', 'POST', wordIndex);
}

/**
 * Gets a list of flashcard words that are due for the user to review
 */
export async function fetchDueWords() {
    return await request('/fetch_due_words', 'GET');
}

/**
 * Creates all flashcards based on words that the user saved using flaskcards_data.json
 */
export async function createInitialCards() {
    return await request('/create_initial_cards', 'POST');
}

/**
 * Creates one card for a word that user has saved
 * 
 * @param {string} word Word to create flashcard for
 */
export async function createCard(word) {
    return await request('/create_card', 'POST', word);
}

/**
 * Updates a card 
 * 
 * @param {string} word Word that card tested user for
 * @param {number} rating Rating of 0, 1, 2, or 3 that user gave word
 */
export async function fetchUpdateCard(word, rating) {
    const data = { word: word, rating: rating };
    
    return await request('update_card', 'POST', data);
}

File: website/app/routes/api/flashcards.py
from datetime import datetime

from flask import Blueprint, request, jsonify

from app.models import db
import app.services.flashcards as flashcards_service


flashcards_bp = Blueprint('flashcards', __name__, url_prefix='/api/v1/flashcards')


@flashcards_bp.post('')
def add_flashcard():
    """Adds a flashcard for the FSRS algorithm"""
    flashcard_data = request.json
    flashcards_service.add_flashcard(db.session, flashcard_data)
    return jsonify('success'), 201


@flashcards_bp.get('/<int:word_id>')
def get_card(word_id: int):
    """Gets a flashcard for the specified word"""
    card = flashcards_service.get_card_for_word_id(db.session, word_id)
    return jsonify(card.to_dict()), 200


@flashcards_bp.get('/next-due')
def get_next_due_flashcard():
    """Gets the next flashcard that is due for review"""
    current_time = request.args.get('current_time')
    flashcard = flashcards_service.get_next_due_flashcard(db.session, current_time)
    if flashcard == 'None':
        return jsonify('None'), 200
    return jsonify(flashcard.to_dict()), 200


@flashcards_bp.get('/due')
def get_due_flashcards():
    """Gets all flashcards that are currently due for review"""
    current_time = request.args.get('current_time')
    flashcard_list = flashcards_service.get_due_flashcards(db.session, current_time)
    return jsonify([flashcard.to_dict() for flashcard in flashcard_list]), 200


@flashcards_bp.patch('/<int:card_id>')
def update_flashcard(card_id: int):
    """Updates a flashcard with its new data"""
    flashcard_data = request.json
    flashcards_service.update_flashcard(db.session, card_id, flashcard_data)
    return jsonify('success'), 200


@flashcards_bp.post('/review')
def review_flashcard():
    """Reviews a flashcard based on the user's rating"""
    review_data = request.json
    word_id = review_data['word_id']
    rating = review_data['rating']
    review_time = datetime.fromisoformat(review_data['review_time'])
    
    card = flashcards_service.get_card_for_word_id(db.session, word_id)
    card = flashcards_service.review_card(card, rating, review_time)
    card_dict = card.to_dict()
    return jsonify(card_dict), 200


@flashcards_bp.get('/review-intervals/<int:card_id>')
def get_review_intervals(card_id: int):
    """Gets the review intervals for a flashcard"""
    review_time = datetime.fromisoformat(request.args.get('review_time'))
    card = flashcards_service.get_card_for_word_id(db.session, card_id)
    print(card_id)
    intervals = flashcards_service.calculate_card_review_intervals(card, review_time)
    return jsonify(intervals), 200


@flashcards_bp.delete('/<int:card_id>')
def delete_flashcard(card_id: int):
    """Deletes a flashcard from the database"""
    flashcards_service.delete_flashcard(db.session, card_id)
    return jsonify('success'), 200